xdem/coreg.py

```diff
@@ -1664,55 +1715,154 @@ def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: Optiona
-        new_nans = nodata_function(y=shifted_east_grid, x=shifted_north_grid)
-        shifted_dem[new_nans >= 1] = np.nan
+def invert_matrix(matrix: np.ndarray) -> np.ndarray:
+    """Invert a transformation matrix."""
+    with warnings.catch_warnings():
+        # Deprecation warning from pytransform3d. Let's hope that is fixed in the near future.
+        warnings.filterwarnings("ignore", message="`np.float` is a deprecated alias for the builtin `float`")
 
-        shifted_dem += self._meta["bias"]
+        # Invert the transform if wanted.
+        return pytransform3d.transformations.invert_transform(matrix)
 
-        return shifted_dem
 
-    def _apply_pts_func(self, coords: np.ndarray) -> np.ndarray:
-        """Apply the estimated x/y/z offsets to a set of points."""
-        offset_east = self._meta["offset_east_px"] * self._meta["resolution"]
-        offset_north = self._meta["offset_north_px"] * self._meta["resolution"]
+def apply_matrix(dem: np.ndarray, transform: rio.transform.Affine, matrix: np.ndarray, invert: bool = False,
+                 centroid: Optional[tuple[float, float, float]] = None,
+                 resampling: Union[int, str] = "cubic") -> np.ndarray:
+    """
+    Apply a transformation matrix to a DEM.
 
-        new_coords = coords.copy()
-        new_coords[:, 0] -= offset_east
-        new_coords[:, 1] -= offset_north
-        new_coords[:, 2] += self._meta["bias"]
+    :param dem: The DEM to transform.
+    :param transform: The Affine transform object (georeferencing) of the DEM.
+    :param matrix: A 4x4 transformation matrix to apply to the DEM.
+    :param invert: Invert the transformation matrix.
+    :param centroid: The X/Y/Z transformation centroid. Irrelevant for pure translations. Defaults to lower left corner.
+    :param resampling: The resampling method to use. Can be `nearest`, `bilinear`, `cubic` or an integer from 0-5.
 
-        return new_coords
+    :returns: The transformed DEM with NaNs as nodata values (replaces a potential mask of the input `dem`).
+    """
+    # Parse the resampling argument given.
+    if isinstance(resampling, int):
+        resampling_order = resampling
+    elif resampling == "cubic":
+        resampling_order = 3
+    elif resampling == "bilinear":
+        resampling_order = 1
+    elif resampling == "nearest":
+        resampling_order = 0
+    else:
+        raise ValueError(
+            f"`{resampling}` is not a valid resampling mode."
+            " Choices: [`nearest`, `bilinear`, `cubic`] or an integer."
+        )
+    # Copy the DEM to make sure the original is not modified, and convert it into an ndarray
+    demc = np.array(dem)
+
+    # Check if the matrix only contains a Z correction. In that case, only shift the DEM values by the bias.
+    empty_matrix = np.diag(np.ones(4, float))
+    matrix_diff = matrix - empty_matrix
+    if abs(matrix_diff[matrix_diff != matrix_diff[2, 3]].mean()) < 1e-4:
```

Comment:
Would that not cause a problem if the matrix has other elements that are equal to element [2, 3]?

Reply:
Also, should it not be strictly equal to 0? The 1e-4 value is somewhat arbitrary here...
In the worse case, we're just doing extra work for nothing.

Reply:
I tried to find a syntax for "slice all values except for this one" but I couldn't find out how to!